### ISSUE TITLE
Using bin edges instead of bin centres

### ISSUE THREAD

**Issue by vandrw** [lucidrains/alphafold3-pytorch#220]
The recent change described in #201 highlighted the need to standardize other areas that currently assumed the use of bin centres. I tried implementing some changes, but I am having a hard time understanding how far the standardization should go. Firstly, I assume that [`ComputeConfidenceScore`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L4529) should be modified to take as input the bins directly instead of the `bin_breaks`, and by extension [`ComputeRankingScore`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L4861), as well as [`ComputeModelSelectionScore`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L5196).  I assume that `bin_breaks` is the same as `distance_bins`? If I understand correctly, the bins would refer to the intervals, while bin breaks are the boundaries? For example, if there are bin breaks at [0, 1, 2, 3], a total of three bins would result: [0, 1), [1, 2), and [2, 3]. The current code initialises `distance_bins` (as well as the other bins) using `torch.linspace(2, 22, 64).float().tolist()`, which is very close to the default initialisation of `pae_breaks = torch.arange(0, 31.5, 0.5)`.

If my assumption is correct, then [`compute_gpde`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L5298) would not require further changes? If that is the case, I could rename `breaks` to `bins` here and in the following, perhaps with a default value of `torch.linspace(0.5, 32, 64)`.
https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L4535-L4536

Getting to the actual modifications to get away from bin centres, I assume we should remove the following function, and modify any subsequent ones that rely on it:
https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L4545

The functions that rely on bin centres seem to be: [`compute_plddt`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L4613), [`compute_ptm`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L4651), and [`compute_pde`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L4726). In a previous discussion, @amorehead mentioned that [`compute_lddt`](https://github.com/lucidrains/alphafold3-pytorch/blob/57d466154ecd3666c035f509fa67971f6323655c/alphafold3_pytorch/alphafold3.py#L5314) (or other lddt related code) might also need to be standardized. However, I cannot seem to find which parts would require attention here.

Please let me know if I am looking at the right code, as I am not entirely sure if I am heading in the right direction.

**Comment by amorehead** [comment]
Hi, @vandrw. First off, thank you for looking into this. Regarding `distance_bins`, when I set its default value to `torch.linspace(2, 22, 64)`, I adopted this value from AF2's supplementary materials, since the AF3 paper points readers to the AF2 paper for the definition of these `distance_bins`. Likewise, I tried to make `dist_breaks` used in `compute_gpde` to match the bins in `distance_bins` as closely as possible (based on the fact that the model's predicted distogram logits are paired with these bins in `ComputeModelSelectionScore`, therefore their bin definitions should line up closely), since as best as I can tell the AF3 supplementary materials do not explicitly state what definition of bins they use for `ComputeModelSelectionScore`. Consequently, I believe `compute_gpde` is already up-to-date and corrected (now).

Regarding the other functions that call `_calculate_bin_centers`, I will leave that up to you whether you think it's necessary to refactor them. At a top-contour glance, they seem correct to me given e.g., their definitions of `pae_breaks` and `pde_breaks`, but perhaps the word `breaks` could simply be changed to `bins` since I think these two terms are functionally synonymous in the current codebase.

When I referred to `plddt` needing to be refactored, I was referring to the end of the `Alphafold3` module's `forward` function (i.e., `plddt_labels = torch.clamp(...)`), where these ground truth labels are computed. I was simply asking for a second opinion on whether this code (aiming to match Section 4.3.1 in the AF3 supplement) could be refactored to use something like `distance_to_dgram`, but perhaps due to its reliance on `lddt_thresholds` it cannot.

**Comment by vandrw** [comment]
Apologies for my confusion! I see now what you meant with the lddt and plddt standardization. From a brief look at the code you referenced and the supplement, everything looks fine!

Regarding the bin centers, I initially assumed that by changing their definition from being the middle of a bin to representing an interval, we would also have to update the calculation of any other labels similarly to maintain consistency. However, I failed to consider that the implementation here might have already taken that into account. Sorry!

If everything checks out, I can close this issue. I mainly wanted to ensure that the previous change didn't require other things to be taken into account.

**Comment by amorehead** [comment]
Thanks again for looking into this. I think the template unit vector function is the last bit of code (I can think of) that could most likely use a bug fix :)
